src/app/rpmostree-compose-builtin-tree.c

```diff
@@ -1056,8 +1227,24 @@ rpmostree_compose_builtin_tree (int             argc,
       return EXIT_FAILURE;
     }
 
-  if (!impl_compose_tree (argv[1], cancellable, error))
+  const char *treefile_path = argv[1];
+
+  g_autoptr(RpmOstreeTreeComposeContext) self = NULL;
+  if (!rpm_ostree_compose_context_new (treefile_path, &self, cancellable, error))
+    return EXIT_FAILURE;
+  gboolean changed;
+  if (!impl_install_tree (self, &changed, cancellable, error))
     return EXIT_FAILURE;
+  if (changed)
```

Comment:
Hmm, I feel like the flow of `impl_install_tree` would be cleaner if we just handled `--dry-run` here too. Though you'd still have to early exit anyway, so meh, maybe not!